internalshared/configutil/kms.go

```diff
@@ -0,0 +1,314 @@
+		// necessarily a string. Then we migrate everything else over to
+		// map[string]string and error if it doesn't work.
+		var m map[string]interface{}
+		if err := hcl.DecodeObject(&m, item.Val); err != nil {
+			return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
+		}
+
+		var purpose []string
+		var err error
+		if v, ok := m["purpose"]; ok {
+			if purpose, err = parseutil.ParseCommaStringSlice(v); err != nil {
+				return multierror.Prefix(fmt.Errorf("unable to parse 'purpose' in kms type %q: %w", key, err), fmt.Sprintf("%s.%s:", blockName, key))
+			}
+			for i, p := range purpose {
+				purpose[i] = strings.ToLower(p)
+			}
+			delete(m, "purpose")
+		}
+
+		strMap := make(map[string]string, len(m))
+		for k, v := range m {
+			if vs, ok := v.(string); ok {
+				strMap[k] = vs
+			} else {
+				return multierror.Prefix(fmt.Errorf("unable to parse 'purpose' in kms type %q: value could not be parsed as string", key), fmt.Sprintf("%s.%s:", blockName, key))
+			}
+		}
+
+		var disabled bool
+		if v, ok := strMap["disabled"]; ok {
+			disabled, err = strconv.ParseBool(v)
+			if err != nil {
+				return multierror.Prefix(err, fmt.Sprintf("%s.%s:", blockName, key))
+			}
+			delete(strMap, "disabled")
+		}
+
+		seal := &KMS{
+			Type:     strings.ToLower(key),
+			Purpose:  purpose,
+			Disabled: disabled,
+		}
+		if len(strMap) > 0 {
+			seal.Config = strMap
+		}
+		seals = append(seals, seal)
+	}
+
+	result.Seals = append(result.Seals, seals...)
+
+	return nil
+}
+
+func configureWrapper(configKMS *KMS, infoKeys *[]string, info *map[string]string, logger hclog.Logger) (wrapping.Wrapper, error) {
+	var wrapper wrapping.Wrapper
+	var kmsInfo map[string]string
+	var err error
+
+	opts := &wrapping.WrapperOptions{
+		Logger: logger,
+	}
+
+	switch configKMS.Type {
+	case wrapping.Shamir:
+		return nil, nil
+
+	case wrapping.AEAD:
+		wrapper, kmsInfo, err = GetAEADKMSFunc(opts, configKMS)
+
+	case wrapping.AliCloudKMS:
+		wrapper, kmsInfo, err = GetAliCloudKMSFunc(opts, configKMS)
+
+	case wrapping.AWSKMS:
+		wrapper, kmsInfo, err = GetAWSKMSFunc(opts, configKMS)
+
+	case wrapping.AzureKeyVault:
+		wrapper, kmsInfo, err = GetAzureKeyVaultKMSFunc(opts, configKMS)
+
+	case wrapping.GCPCKMS:
+		wrapper, kmsInfo, err = GetGCPCKMSKMSFunc(opts, configKMS)
+
+	case wrapping.OCIKMS:
+		wrapper, kmsInfo, err = GetOCIKMSKMSFunc(opts, configKMS)
+
+	case wrapping.Transit:
+		wrapper, kmsInfo, err = GetTransitKMSFunc(opts, configKMS)
+
+	case wrapping.PKCS11:
+		return nil, fmt.Errorf("KMS type 'pkcs11' requires the Vault Enterprise HSM binary")
+
+	default:
+		return nil, fmt.Errorf("Unknown KMS type %q", configKMS.Type)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range kmsInfo {
+		*infoKeys = append(*infoKeys, k)
+		(*info)[k] = v
+	}
+
+	return wrapper, nil
+}
+
+var GetAEADKMSFunc = func(opts *wrapping.WrapperOptions, kms *KMS) (wrapping.Wrapper, map[string]string, error) {
```

Comment:
Why are these declared as vars as opposed to being functions directly?

Reply:
An excellent question for which I cannot remember the answer. I've changed them back to funcs.